File: packages/flutter_blue_plus_android/android/src/main/kotlin/com/lib/flutter_blue_plus/PendingOperations.kt
// Copyright 2017-2023, Charles Weinberger & Paul DeMarco.
// All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

package com.lib.flutter_blue_plus

import android.os.Handler

/**
 * The snake_case wire form of this error code (crosses the channel as
 * `PlatformException.code`).
 */
val FbpErrorCode.wire: String get() = name.lowercase()

/**
 * Identifies one in-flight request awaiting completion from a native
 * callback (GATT callback, broadcast receiver, activity result, ...).
 */
sealed class OpKey {
    abstract val address: String?

    data class Connect(override val address: String) : OpKey()
    data class Disconnect(override val address: String) : OpKey()
    data class DiscoverServices(override val address: String) : OpKey()
    data class ReadChar(
        override val address: String,
        val svcUuid: String,
        val svcInstance: Int,
        val chrUuid: String,
        val chrInstance: Int,
    ) : OpKey()
    data class WriteChar(
        override val address: String,
        val svcUuid: String,
        val svcInstance: Int,
        val chrUuid: String,
        val chrInstance: Int,
    ) : OpKey()
    data class ReadDesc(
        override val address: String,
        val svcUuid: String,
        val svcInstance: Int,
        val chrUuid: String,
        val chrInstance: Int,
        val descUuid: String,
    ) : OpKey()
    data class WriteDesc(
        override val address: String,
        val svcUuid: String,
        val svcInstance: Int,
        val chrUuid: String,
        val chrInstance: Int,
        val descUuid: String,
    ) : OpKey()
    data class SetNotify(
        override val address: String,
        val svcUuid: String,
        val svcInstance: Int,
        val chrUuid: String,
        val chrInstance: Int,
    ) : OpKey()
    data class Mtu(override val address: String) : OpKey()
    data class Rssi(override val address: String) : OpKey()
    data class Phy(override val address: String) : OpKey()
    data class CreateBond(override val address: String) : OpKey()
    data class RemoveBond(override val address: String) : OpKey()
    data class ClearCache(override val address: String) : OpKey()
    object TurnOn : OpKey() {
        override val address: String? get() = null
    }
}

/**
 * Registry correlating async pigeon completion callbacks with the native
 * callbacks that eventually complete them.
 *
 * All completions are dispatched exactly once, on the main looper.
 */
class PendingOperations(private val mainHandler: Handler) {

    private val ops = HashMap<OpKey, (Result<Any?>) -> Unit>()

    /**
     * Registers [callback] for [key]. If an operation with the same key is
     * already in flight, the NEW callback is failed immediately with
     * "operation_in_progress" and the existing one is left untouched.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T> register(key: OpKey, callback: (Result<T>) -> Unit): Boolean {
        val cb = callback as (Result<Any?>) -> Unit
        val conflict = synchronized(ops) {
            if (ops.containsKey(key)) {
                true
            } else {
                ops[key] = cb
                false
            }
        }
        if (conflict) {
            dispatch(cb, Result.failure(FlutterError(
                FbpErrorCode.OPERATION_IN_PROGRESS.wire,
                "an operation of this type is already in progress for this device", null)))
        }
        return !conflict
    }

    /** Atomic remove-and-return. Guarantees exactly-once completion. */
    fun take(key: OpKey): ((Result<Any?>) -> Unit)? {
        return synchronized(ops) { ops.remove(key) }
    }

    /**
     * Completes the operation for [key] with [result], if one is pending.
     * Returns whether an operation was pending.
     */
    fun complete(key: OpKey, result: Result<Any?>): Boolean {
        val cb = take(key) ?: return false
        dispatch(cb, result)
        return true
    }

    /** Completes the operation for [key] with [value], if one is pending. */
    fun succeed(key: OpKey, value: Any?): Boolean = complete(key, Result.success(value))

    /** Fails the operation for [key] with [error], if one is pending. */
    fun fail(key: OpKey, error: FlutterError): Boolean = complete(key, Result.failure(error))

    /** Fails every pending operation belonging to [address]. */
    fun failAllForDevice(address: String, error: FlutterError) {
        val taken = synchronized(ops) {
            val keys = ops.keys.filter { it.address == address }
            keys.mapNotNull { ops.remove(it) }
        }
        taken.forEach { dispatch(it, Result.failure(error)) }
    }

    /** Fails every pending operation. */
    fun failAll(error: FlutterError) {
        val taken = synchronized(ops) {
            val callbacks = ops.values.toList()
            ops.clear()
            callbacks
        }
        taken.forEach { dispatch(it, Result.failure(error)) }
    }

    /** Drops all pending operations WITHOUT invoking them (hot restart). */
    fun clearAll() {
        synchronized(ops) { ops.clear() }
    }

    private fun dispatch(cb: (Result<Any?>) -> Unit, result: Result<Any?>) {
        mainHandler.post { cb(result) }
    }
}

File: packages/flutter_blue_plus_darwin/darwin/flutter_blue_plus_darwin/Sources/flutter_blue_plus_darwin/PendingOperations.swift
// Copyright 2017-2023, Charles Weinberger & Paul DeMarco.
// All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import Foundation

#if os(iOS)
  import Flutter
#elseif os(macOS)
  import FlutterMacOS
#endif

/// Key for pending operations targeting a specific attribute of a device.
struct DeviceOpKey<Ref: Hashable>: Hashable {
  let address: String
  let ref: Ref
}

/// A map of in-flight completions for a single kind of operation.
///
/// Contract:
/// - `register` fails the *new* completion with "operation_in_progress" if an
///   operation for the same key already exists, and returns false.
/// - `take` removes and returns the completion, guaranteeing exactly-once
///   invocation by the caller.
final class PendingMap<Key: Hashable, Value> {
  private var completions: [Key: (Result<Value, Error>) -> Void] = [:]

  /// Registers a completion. Returns false (and fails `completion`) if an
  /// operation with the same key is already in flight.
  func register(_ key: Key, _ completion: @escaping (Result<Value, Error>) -> Void) -> Bool {
    if completions[key] != nil {
      completion(
        .failure(
          PigeonError(
            code: FbpErrorCode.operationInProgress.wire,
            message: "this operation is already in progress",
            details: nil)))
      return false
    }
    completions[key] = completion
    return true
  }

  /// Removes and returns the completion for `key`, if any.
  func take(_ key: Key) -> ((Result<Value, Error>) -> Void)? {
    return completions.removeValue(forKey: key)
  }

  /// Fails every pending operation whose key matches `predicate`.
  func failAll(where predicate: (Key) -> Bool, error: Error) {
    let keys = completions.keys.filter(predicate)
    for key in keys {
      completions.removeValue(forKey: key)?(.failure(error))
    }
  }

  /// Fails every pending operation.
  func failAll(error: Error) {
    let all = completions
    completions.removeAll()
    for (_, completion) in all {
      completion(.failure(error))
    }
  }

  /// Drops all pending operations *without* invoking them
  /// (hot restart / flutterRestart — the dart side no longer exists).
  func clearAll() {
    completions.removeAll()
  }
}

/// Typed per-operation storage for all @async host-api calls that complete
/// from CoreBluetooth delegate callbacks.
final class PendingOperations {
  let connect = PendingMap<String, Void>()
  let disconnect = PendingMap<String, Void>()
  let discover = PendingMap<String, [BmBluetoothService]>()
  let rssi = PendingMap<String, Int64>()
  let charRead = PendingMap<DeviceOpKey<BmCharacteristicRef>, FlutterStandardTypedData>()
  let charWrite = PendingMap<DeviceOpKey<BmCharacteristicRef>, Void>()
  let setNotify = PendingMap<DeviceOpKey<BmCharacteristicRef>, Bool>()
  let descRead = PendingMap<DeviceOpKey<BmDescriptorRef>, FlutterStandardTypedData>()
  let descWrite = PendingMap<DeviceOpKey<BmDescriptorRef>, Void>()

  /// Fails every attribute-level operation for one device (used when the
  /// device disconnects). Connect/disconnect are handled explicitly by the
  /// caller before this is invoked.
  func failAllForDevice(_ address: String, error: Error) {
    discover.failAll(where: { $0 == address }, error: error)
    rssi.failAll(where: { $0 == address }, error: error)
    charRead.failAll(where: { $0.address == address }, error: error)
    charWrite.failAll(where: { $0.address == address }, error: error)
    setNotify.failAll(where: { $0.address == address }, error: error)
    descRead.failAll(where: { $0.address == address }, error: error)
    descWrite.failAll(where: { $0.address == address }, error: error)
  }

  /// Fails every pending operation of every kind (adapter turned off).
  func failAll(error: Error) {
    connect.failAll(error: error)
    disconnect.failAll(error: error)
    discover.failAll(error: error)
    rssi.failAll(error: error)
    charRead.failAll(error: error)
    charWrite.failAll(error: error)
    setNotify.failAll(error: error)
    descRead.failAll(error: error)
    descWrite.failAll(error: error)
  }

  /// Drops everything without invoking (hot restart / engine detach).
  func clearAll() {
    connect.clearAll()
    disconnect.clearAll()
    discover.clearAll()
    rssi.clearAll()
    charRead.clearAll()
    charWrite.clearAll()
    setNotify.clearAll()
    descRead.clearAll()
    descWrite.clearAll()
  }
}
